contracts/cw721-stake/src/contract.rs

```diff
@@ -0,0 +1,527 @@
+use crate::hooks::{stake_hook_msgs, unstake_hook_msgs};
+#[cfg(not(feature = "library"))]
+use crate::msg::{
+    ExecuteMsg, GetHooksResponse, InstantiateMsg, Owner, QueryMsg, StakedBalanceAtHeightResponse,
+    TotalStakedAtHeightResponse,
+};
+use crate::state::{
+    Config, CONFIG, HOOKS, MAX_CLAIMS, NFT_CLAIMS, STAKED_NFTS_PER_OWNER, TOTAL_STAKED_NFTS,
+};
+use crate::ContractError;
+use cosmwasm_std::{
+    entry_point, to_binary, Addr, Binary, CosmosMsg, Deps, DepsMut, Empty, Env, MessageInfo,
+    Response, StdError, StdResult, Uint128, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw721::Cw721ReceiveMsg;
+use cw_utils::Duration;
+use indexmap::IndexSet;
+use std::convert::{From, TryFrom};
+
+pub(crate) const CONTRACT_NAME: &str = "crates.io:cw721_stake";
+pub(crate) const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response<Empty>, ContractError> {
+    let owner = msg
+        .owner
+        .as_ref()
+        .map(|owner| match owner {
+            Owner::Addr(address) => deps.api.addr_validate(address),
+            Owner::Instantiator {} => Ok(info.sender),
+        })
+        .transpose()?;
+    let manager = msg
+        .manager
+        .as_ref()
+        .map(|h| deps.api.addr_validate(h))
+        .transpose()?;
+
+    let config = Config {
+        owner: owner.clone(),
+        manager,
+        nft_address: deps.api.addr_validate(&msg.nft_address)?,
```

Comment:
so basic question, why do we only allow staking one kind of NFT?

Reply:
Such a good question! We could without too much trouble. Would just worry that this would require more thinking about how to assign weights between the two collections and etc. which might be left for later. @Callum-A's voting module weighted combiner will give this to us for free as well.

Reply:
> Such a good question! We could without too much trouble. Would just worry that this would require more thinking about how to assign weights between the two collections and etc. which might be left for later. @Callum-A's voting module weighted combiner will give this to us for free as well.

Exactly!